wpilib/wpilib/pidbase.py

```diff
@@ -0,0 +1,500 @@
+# validated: 2018-01-27 DS 738a1c015cfb edu/wpi/first/wpilibj/PIDBase.java
+# ----------------------------------------------------------------------------
+# Copyright (c) FIRST 2008-2018. All Rights Reserved.
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+# ----------------------------------------------------------------------------
+
+import threading
+import warnings
+
+import hal
+
+from .interfaces import PIDSource
+from .sendablebase import SendableBase
+from .resource import Resource
+from .timer import Timer
+from .lineardigitalfilter import LinearDigitalFilter
+
+__all__ = ["PIDBase"]
+
+
+class PIDBase(SendableBase):
+    """Class implements a PID Control Loop.
+
+    Creates a separate thread which reads the given PIDSource and takes care of the integral
+    calculations, as well as writing the given PIDOutput.
+
+    This feedback controller runs in discrete time, so time deltas are not used in the integral
+    and derivative calculations. Therefore, the sample rate affects the controller's behavior for a
+    given set of PID constants.
+
+    """
+
+    kDefaultPeriod = 0.05
+    instances = 0
+
+    PIDSourceType = PIDSource.PIDSourceType
+
+    # Tolerance is the type of tolerance used to specify if the PID controller
+    # is on target.  The various implementations of this such as
+    # PercentageTolerance and AbsoluteTolerance specify types of tolerance
+    # specifications to use.
+    def PercentageTolerance_onTarget(self, percentage):
+        return abs(self.getError()) < percentage / 100.0 * self.inputRange
+
+    def AbsoluteTolerance_onTarget(self, value):
+        return abs(self.getError()) < value
+
+    def __init__(self, Kp, Ki, Kd, Kf, output, source):
```

Comment:
Java has a default for Kf here. Do you think it's worth putting it here too? We can make source and output keyword-only arguments.

Wait, source and output are in the wrong order compared to Java too.

Reply:
I think it's not worth putting on the base

Reply:
The argument order still needs to be fixed though.

Having the default here could simplify the PIDController constructor too.